test/epgsql_replication_tests.erl

```diff
@@ -0,0 +1,180 @@
+                  false ->
+                      error_logger:info_msg("Git version of test schema does not match: ~p ~p~nPlease run make create_testdbs to update your test databases", [Version, DBVersion]),
+                      erlang:exit(1);
+                  _ ->
+                      undefined
+              end
+      end),
+
+    Tests =
+        lists:map(
+          fun({Name, _}) ->
+                  {Name, fun(X) -> ?MODULE:Name(X) end}
+          end,
+          lists:filter(
+            fun({Name, Arity}) ->
+                    case {lists:suffix("_test", atom_to_list(Name)), Arity} of
+                        {true, 1} -> true;
+                        _ -> false
+                    end
+            end,
+            ?MODULE:module_info(functions))),
+    WithModule =
+        fun(Module) ->
+                lists:map(
+                  fun({Name, Test}) ->
+                          {lists:flatten(
+                             io_lib:format("~s(~s)", [Name, Module])),
+                           fun() -> Test(Module) end}
+                  end,
+                  Tests)
+        end,
+    [WithModule(epgsql)
+    ].
+
+%% -- internal functions --
+
+connect_only(Module, Args) ->
+    TestOpts = [{port, ?port}],
+    case Args of
+        [User, Opts]       -> Args2 = [User, TestOpts ++ Opts];
+        [User, Pass, Opts] -> Args2 = [User, Pass, TestOpts ++ Opts];
+        Opts               -> Args2 = [TestOpts ++ Opts]
+    end,
+    {ok, C} = apply(Module, connect, [?host | Args2]),
+    Module:close(C),
+    flush().
+
+with_connection(Module, F) ->
+    with_connection(Module, F, "epgsql_test", []).
+
+with_connection(Module, F, Args) ->
+    with_connection(Module, F, "epgsql_test", Args).
+
+with_connection(Module, F, Username, Args) ->
+    Args2 = [{port, ?port}, {database, "epgsql_test_db1"} | Args],
+    {ok, C} = Module:connect(?host, Username, Args2),
+    try
+        F(C)
+    after
+        Module:close(C)
+    end,
+    flush().
+
+%% flush mailbox
+flush() ->
+    ?assertEqual([], flush([])).
+
+flush(Acc) ->
+    receive
+        {'EXIT', _Pid, normal} -> flush(Acc);
+        M                      -> flush([M | Acc])
+    after
+        0 -> lists:reverse(Acc)
+    end.
+
+replication_test_run(Module, Callback) ->
+    with_connection(
+        Module,
+        fun(C) ->
+            {ok, _, _} = Module:squery(C, "CREATE_REPLICATION_SLOT ""epgsql_test"" LOGICAL ""test_decoding"""),
+
+            %% new connection because main id in a replication mode
+            with_connection(
+                Module,
+                fun(C2) ->
+                    [{ok, 1},{ok, 1}] = Module:squery(C2,
+                        "insert into test_table1 (id, value) values (5, 'five');delete from test_table1 where id = 5;")
+                end,
+                "epgsql_test_db1"),
+
+            Module:start_replication(C, "epgsql_test", Callback, #{conn => C, pid => self()}, "0/0"),
```

Comment:
Maps are not yet supported on all Erlang versions supported by epgsql. This will trigger travis fall.